File: Array/zhangjianan/20200616有效括号/Solution.java
package com.test;

import java.util.HashMap;
import java.util.Stack;

public class Solution {
   static HashMap<Character, Character> map = new HashMap<>();

    public static void main(String[] args) {
        System.out.println(isValid("()"));
        System.out.println(isValid("(({}))"));
        System.out.println(isValid("(({}{{})})"));
        System.out.println(isValid("("));
    }

    private static boolean isValid(String s) {
        map.put(')','(');
        map.put(']','[');
        map.put('}','{');
        char[] chars = s.toCharArray();
		 if(chars.length%2==1)return false;
        Stack<Character> stack=new Stack<Character>();
        for (int i = 0; i < chars.length; i++) {
            if (map.containsKey(chars[i])) {
                Character character = map.get(chars[i]);
                Character pop = stack.empty() ? '#' : stack.pop();
                if (pop != character) {
                    return false;
                }
            } else {
                stack.push(chars[i]);
            }
        }
        return stack.empty();
    }
}

File: Array/zhangjianan/20200616有效括号/Solution.kt
package com.test

import java.util.HashMap
import java.util.Stack

object Solution {
    internal var map = HashMap<Char, Char>()

    @JvmStatic
    fun main(args: Array<String>) {
        println(isValid("()"))
        println(isValid("(({}))"))
        println(isValid("(({}{{})})"))
        println(isValid("("))
    }

    private fun isValid(s: String): Boolean {
        map.put(')','(')
        map.put(']','[')
        map.put('}','{')
        val chars=s.toCharArray()
        if(chars.size%2==1) return false
        val stack=Stack<Char>()
        for (i in 0 until chars.size){
            if (map.containsKey(chars[i])) {
                val character = map.get(chars[i])
                var pop = if(stack.empty()) {'#'}  else{stack.pop()}
                if (pop != character) {
                    return false
                }
            } else {
                stack.push(chars[i])
            }
        }
        return stack.empty()
    }
}
